File: typescript/src/auditBundle.ts
/**
 * Offline verification for audit export bundles.
 *
 * Mirrors `atlasent-api/supabase/functions/v1-audit/verify.ts`. The
 * reference verifier there is the source of truth; this module must
 * stay byte-identical with it on the canonicalization + signing path
 * so a bundle that verifies in the backend verifies here (and vice
 * versa).
 *
 * Primary entry point:
 *
 * ```ts
 * import { verifyBundle } from "@atlasent/sdk";
 * const result = await verifyBundle("export.json", { publicKeysPem: [pem] });
 * ```
 *
 * Node 20+ ships Ed25519 in `crypto.webcrypto.subtle`, so no extra
 * dependencies are required.
 */
import { readFile } from "node:fs/promises";
import { webcrypto } from "node:crypto";
import type { TrustRootSnapshot } from "./trustRoot.js";
import { getGlobalTrustRootManager } from "./trustRoot.js";
import { BundleVerificationError } from "./errors.js";

const GENESIS_HASH = "0".repeat(64);

const subtle = webcrypto.subtle;

/** Node's webcrypto CryptoKey — kept local so the module doesn't depend on DOM types. */
type WebCryptoKey = webcrypto.CryptoKey;

/** Public key candidate the verifier will try, tagged with its registry id. */
export interface VerifyKey {
  keyId: string;
  publicKey: WebCryptoKey;
  /**
   * Raw 32-byte Ed25519 public key, when known. Populated automatically for
   * keys loaded from `publicKeysPem`. Lets the ADR-005 revocation / role
   * check identify the key that ACTUALLY verified the signature in the trust
   * root by material (`TrustRootKey.x`), instead of trusting the bundle's
   * unsigned `signing_key_id` hint — a bundle signed by a revoked key but
   * advertising a live kid must still fail (Codex P1 on atlasent-sdk#519).
   * `publicKey` is imported non-extractable, so the material is captured
   * here at load time rather than exported later.
   */
  publicKeyRaw?: Uint8Array;
}

export interface BundleVerificationResult {
  /**
   * AND of three checks: adjacency (each event's `previous_hash`
   * equals the prior event's `hash`), per-event hash recomputation
   * from the canonical payload, and `chain_head_hash` matching the
   * last event's stored hash.
   */
  chainIntegrityOk: boolean;
  /** Ed25519 signature verified against one of the supplied public keys. */
  signatureValid: boolean;
  /** `chain_head_hash` equals the last event's stored `hash`. */
  headHashMatches: boolean;
  /** Event ids whose recomputed hash != stored hash. */
  tamperedEventIds: string[];
  /** Which registry key id matched, when `signatureValid` is true. */
  matchedKeyId?: string | undefined;
  /** Non-fatal explanation when a flag is false. */
  reason?: string | undefined;
  /** Convenience: `chainIntegrityOk && signatureValid`. */
  verified: boolean;
}

/** Parsed bundle shape the verifier consumes. Fields beyond these are ignored. */
export interface AuditBundle {
  export_id?: unknown;
  org_id?: unknown;
  chain_head_hash?: unknown;
  event_count?: unknown;
  signed_at?: unknown;
  events?: unknown;
  signature?: unknown;
  signing_key_id?: unknown;
  [k: string]: unknown;
}

export interface VerifyBundleOptions {
  /** SPKI-PEM strings (one per key in the active trust set). */
  publicKeysPem?: readonly string[];
  /** Already-imported keys, paired with registry ids (rotation hint). */
  keys?: readonly VerifyKey[];
  /** Trust-root snapshot for revocation + expiry checks. */
  trustRoot?: TrustRootSnapshot;
  /**
   * Opt out of fail-closed snapshot expiry check (ADR-005 D3).
   * Intended for air-gap / offline use cases.
   * Emits a one-time warning per process start.
   */
  allowExpiredSnapshot?: boolean;
}

// ─── Canonicalization ─────────────────────────────────────────────────────────

/**
 * Reproduces `_shared/rules.ts::canonicalJSON` byte-for-byte:
 *   - object keys sorted at every depth
 *   - no whitespace
 *   - `null`, `undefined`, `NaN`, `±Infinity` all render as `"null"`
 *   - strings use standard `JSON.stringify` escapes
 */
export function canonicalJSON(value: unknown): string {
  if (value === null || value === undefined) return "null";
  if (typeof value === "number") return Number.isFinite(value) ? String(value) : "null";
  if (typeof value === "boolean") return value ? "true" : "false";
  if (typeof value === "string") return JSON.stringify(value);
  if (Array.isArray(value)) return "[" + value.map(canonicalJSON).join(",") + "]";
  if (typeof value === "object") {
    const obj = value as Record<string, unknown>;
    const keys = Object.keys(obj).sort();
    return "{" + keys.map((k) => JSON.stringify(k) + ":" + canonicalJSON(obj[k])).join(",") + "}";
  }
  return "null";
}

async function sha256Hex(input: string): Promise<string> {
  const buf = await subtle.digest("SHA-256", new TextEncoder().encode(input));
  return Array.from(new Uint8Array(buf))
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

// ─── Envelope reconstruction ──────────────────────────────────────────────────

/**
 * Recreate the exact bytes `handleExport` signed. Key order is
 * load-bearing — must match the object literal in
 * `v1-audit/index.ts::handleExport`. V8 preserves insertion order, so
 * the literal below is byte-identical with what the backend signs.
 */
export function signedBytesFor(bundle: AuditBundle): Uint8Array<ArrayBuffer> {
  const envelope = {
    export_id: bundle.export_id,
    org_id: bundle.org_id,
    chain_head_hash: bundle.chain_head_hash,
    event_count: bundle.event_count,
    signed_at: bundle.signed_at,
    events: bundle.events,
  };
  return new TextEncoder().encode(JSON.stringify(envelope));
}

// ─── Chain verification ───────────────────────────────────────────────────────

interface ChainEvent {
  id?: unknown;
  hash?: unknown;
  previous_hash?: unknown;
  payload?: unknown;
}

async function verifyChainEvents(
  events: ChainEvent[],
): Promise<{ adjacencyOk: boolean; tamperedIds: string[] }> {
  const tamperedIds: string[] = [];
  let adjacencyOk = true;
  const first = events[0];
  let prevHash =
    first && typeof first.previous_hash === "string" ? first.previous_hash : GENESIS_HASH;

  for (let i = 0; i < events.length; i++) {
    const e = events[i];
    if (!e || typeof e.hash !== "string" || typeof e.previous_hash !== "string") {
      tamperedIds.push(String(e?.id ?? `index_${i}`));
      adjacencyOk = false;
      continue;
    }
    if (e.previous_hash !== prevHash) adjacencyOk = false;

    const canonical = canonicalJSON(e.payload ?? {});
    const recomputed = await sha256Hex(prevHash + canonical);
    if (recomputed !== e.hash) tamperedIds.push(String(e.id));

    prevHash = e.hash;
  }

  return { adjacencyOk, tamperedIds };
}

// ─── Signature verification ───────────────────────────────────────────────────

function base64UrlDecode(s: string): Uint8Array<ArrayBuffer> {
  const pad = s.length % 4 === 0 ? "" : "=".repeat(4 - (s.length % 4));
  const b64 = s.replace(/-/g, "+").replace(/_/g, "/") + pad;
  const bin = Buffer.from(b64, "base64");
  const out = new Uint8Array(bin.byteLength);
  out.set(bin);
  return out;
}

/** DER prefix of an Ed25519 SubjectPublicKeyInfo (RFC 8410): 12 bytes, then the 32-byte raw key. */
const ED25519_SPKI_PREFIX = Uint8Array.from([
  0x30, 0x2a, 0x30, 0x05, 0x06, 0x03, 0x2b, 0x65, 0x70, 0x03, 0x21, 0x00,
]);

/** Extract the raw 32-byte key from an Ed25519 SPKI DER, or undefined if the DER is not that exact shape. */
export function rawEd25519FromSpki(spki: Uint8Array): Uint8Array | undefined {
  if (spki.length !== ED25519_SPKI_PREFIX.length + 32) return undefined;
  for (let i = 0; i < ED25519_SPKI_PREFIX.length; i++) {
    if (spki[i] !== ED25519_SPKI_PREFIX[i]) return undefined;
  }
  return spki.slice(ED25519_SPKI_PREFIX.length);
}

function base64UrlEncode(bytes: Uint8Array): string {
  // Node ≥ 16 emits base64url natively; no padding, URL-safe alphabet. This
  // avoids a trailing-`=` strip regex (CodeQL flags `=+$` as polynomial on
  // untrusted input, and the bytes here can come from a caller's key).
  return Buffer.from(bytes).toString("base64url");
}

/** Import an Ed25519 SPKI PEM as a `VerifyKey`, capturing the raw material for trust-root matching. */
export async function verifyKeyFromSpkiPem(pem: string, keyId: string): Promise<VerifyKey> {
  const b64 = pem
    .replace(/-----BEGIN PUBLIC KEY-----/, "")
    .replace(/-----END PUBLIC KEY-----/, "")
    .replace(/\s+/g, "");
  const bytes = Uint8Array.from(Buffer.from(b64, "base64"));
  const publicKey = await subtle.importKey("spki", bytes, { name: "Ed25519" }, false, ["verify"]);
  const publicKeyRaw = rawEd25519FromSpki(bytes);
  return publicKeyRaw ? { keyId, publicKey, publicKeyRaw } : { keyId, publicKey };
}

async function resolveKeys(options: VerifyBundleOptions | undefined): Promise<VerifyKey[]> {
  const out: VerifyKey[] = [];
  if (options?.keys) out.push(...options.keys);
  if (options?.publicKeysPem) {
    for (let i = 0; i < options.publicKeysPem.length; i++) {
      const pem = options.publicKeysPem[i];
      if (!pem) continue;
      try {
        out.push(await verifyKeyFromSpkiPem(pem, `pem_${i}`));
      } catch {
        // Malformed PEM — skip it, try the rest.
      }
    }
  }
  return out;
}

// ─── Public API ───────────────────────────────────────────────────────────────

export async function verifyAuditBundle(
  bundle: AuditBundle,
  keys: readonly VerifyKey[],
  trustRootOpts?: {
    trustRoot?: TrustRootSnapshot;
    allowExpiredSnapshot?: boolean;
  },
): Promise<BundleVerificationResult> {
  // ── ADR-005 D3: fail-closed snapshot expiry check ──────────────────
  if (trustRootOpts?.trustRoot) {
    const snap = trustRootOpts.trustRoot;
    const now = Date.now();
    const validUntil = new Date(snap.valid_until).getTime();
    if (now > validUntil && !trustRootOpts.allowExpiredSnapshot) {
      throw new BundleVerificationError({
        reason: "trust_snapshot_expired",
        snapshotValidUntil: snap.valid_until,
        snapshotFetchedAt: snap.issued_at,
      });
    }
  }

  const events: ChainEvent[] = Array.isArray(bundle.events) ? (bundle.events as ChainEvent[]) : [];

  const { adjacencyOk, tamperedIds } = await verifyChainEvents(events);

  const last = events[events.length - 1];
  const lastHash = last && typeof last.hash === "string" ? last.hash : GENESIS_HASH;
  const headHashMatches =
    typeof bundle.chain_head_hash === "string" ? bundle.chain_head_hash === lastHash : false;

  const chainIntegrityOk = adjacencyOk && tamperedIds.length === 0 && headHashMatches;

  let signatureValid = false;
  let matchedKeyId: string | undefined;
  let matchedKeyRaw: Uint8Array | undefined;
  let reason: string | undefined;

  if (keys.length === 0) {
    reason =
      "no signing keys configured (signing_keys table empty and ATLASENT_EXPORT_SIGNING_KEY_PUBLIC unset)";
  } else if (typeof bundle.signature !== "string" || bundle.signature.length === 0) {
    reason = "bundle carries no signature";
  } else {
    try {
      const sigBytes = base64UrlDecode(bundle.signature);
      const envelopeBytes = signedBytesFor(bundle);
      const hint = typeof bundle.signing_key_id === "string" ? bundle.signing_key_id : null;
      const ordered = hint
        ? [
            ...keys.filter((k) => k.keyId === hint),
            ...keys.filter((k) => k.keyId !== hint),
          ]
        : Array.from(keys);
      for (const k of ordered) {
        const ok = await subtle.verify("Ed25519", k.publicKey, sigBytes, envelopeBytes);
        if (ok) {
          signatureValid = true;
          matchedKeyId = k.keyId;
          matchedKeyRaw = k.publicKeyRaw;
          break;
        }
      }
      if (!signatureValid) {
        reason = `signature did not verify under any of ${keys.length} configured public key(s)`;
      }
    } catch (err) {
      reason = `signature check failed: ${err instanceof Error ? err.message : String(err)}`;
    }
  }

  // ── ADR-005: revocation + role check (after signature, before returning) ──
  //
  // `signing_key_id` is OUTSIDE the signed envelope (see signedBytesFor), so it
  // is a hint the producer — or an attacker holding a revoked key — chooses
  // freely. The checks below are therefore anchored to the key that ACTUALLY
  // verified the signature, identified in the trust root by raw material
  // (`TrustRootKey.x`) whenever the verifying key carries it; the hint is used
  // only to disambiguate trust-root entries that share one material, and is
  // additionally rejected on its own if it names a revoked kid (fail-closed —
  // a bundle that advertises a revoked key is wrong either way). Keys supplied
  // without material (`VerifyKey.publicKeyRaw` absent) fall back to the
  // hint-only behaviour, which is exactly the pre-2026-09-13 semantics.
  if (signatureValid && trustRootOpts?.trustRoot) {
    const snap = trustRootOpts.trustRoot;
    const hint = typeof bundle.signing_key_id === "string" ? bundle.signing_key_id : null;
    const matchedX = matchedKeyRaw ? base64UrlEncode(matchedKeyRaw) : null;
    const byMaterial = matchedX ? snap.keys.filter((k) => k.x === matchedX) : [];
    const byId = matchedKeyId ? snap.keys.filter((k) => k.kid === matchedKeyId) : [];
    const sameMaterial = byMaterial.length > 0 ? byMaterial : byId;
    const hintEntry = hint ? sameMaterial.find((k) => k.kid === hint) : undefined;
    // The trust-root entries that describe the verifying key. If the hint
    // names one of them it is authoritative; otherwise every entry with that
    // material counts (a revocation of the material under any kid revokes it).
    const verifyingEntries = hintEntry ? [hintEntry] : sameMaterial;

    const revokedKid = [
      ...verifyingEntries.map((k) => k.kid),
      ...(matchedKeyId ? [matchedKeyId] : []),
      ...(hint ? [hint] : []),
    ].find((kid) => snap.revoked_keys.some((r) => r.kid === kid));
    if (revokedKid !== undefined) {
      throw new BundleVerificationError({
        reason: "key_revoked",
        snapshotValidUntil: snap.valid_until,
        snapshotFetchedAt: snap.issued_at,
        kid: revokedKid,
      });
    }
    // Role: audit bundles must be signed by an R3_audit key. Judge the
    // verifying key's own entry; fall back to the hint's entry when the
    // verifying key is not in the trust root at all (customer-supplied PEM).
    const roleEntry =
      verifyingEntries.find((k) => k.role !== "R3_audit") ??
      (verifyingEntries.length === 0 && hint ? snap.keys.find((k) => k.kid === hint) : undefined);
    if (roleEntry && roleEntry.role !== "R3_audit") {
      throw new BundleVerificationError({
        reason: "key_role_mismatch",
        snapshotValidUntil: snap.valid_until,
        snapshotFetchedAt: snap.issued_at,
        kid: roleEntry.kid,
      });
    }
  }

  if (!chainIntegrityOk && reason === undefined) {
    if (tamperedIds.length > 0) reason = `hash mismatch for ${tamperedIds.length} event(s)`;
    else if (!adjacencyOk) reason = "chain adjacency broken";
    else if (!headHashMatches) reason = "chain_head_hash does not match last event";
  }

  return {
    chainIntegrityOk,
    signatureValid,
    headHashMatches,
    tamperedEventIds: tamperedIds,
    matchedKeyId,
    reason,
    verified: chainIntegrityOk && signatureValid,
  };
}

/**
 * Load a bundle from disk (or a parsed object) and verify it.
 *
 * `publicKeysPem` is the active SPKI-PEM set from
 * `GET /v1-signing-keys`. When omitted, the chain check still runs
 * but `signatureValid` will be false with an explanatory `reason` —
 * callers that want a complete offline check MUST supply the trust
 * set.
 *
 * When `trustRoot` is not supplied, the global trust-root manager's
 * current snapshot is used automatically (B2.3 bootstrap wire-in).
 * Pass `allowExpiredSnapshot: true` to disable fail-closed expiry
 * for air-gap environments.
 */
export async function verifyBundle(
  pathOrBundle: string | AuditBundle,
  options?: VerifyBundleOptions,
): Promise<BundleVerificationResult> {
  let bundle: AuditBundle;
  if (typeof pathOrBundle === "string") {
    const raw = await readFile(pathOrBundle, "utf8");
    const parsed = JSON.parse(raw);
    // Fixture wrapper shape: { description, bundle }. Accepted for ergonomics.
    bundle =
      parsed && typeof parsed === "object" && "bundle" in parsed && typeof parsed.bundle === "object"
        ? (parsed.bundle as AuditBundle)
        : (parsed as AuditBundle);
  } else {
    bundle = pathOrBundle;
  }
  const keys = await resolveKeys(options);
  // Auto-inject the global trust-root snapshot when none is explicitly provided.
  const effectiveTrustRoot =
    options?.trustRoot ?? getGlobalTrustRootManager().getSnapshot();
  return verifyAuditBundle(bundle, keys, {
    trustRoot: effectiveTrustRoot,
    ...(options?.allowExpiredSnapshot !== undefined && {
      allowExpiredSnapshot: options.allowExpiredSnapshot,
    }),
  });
}

File: python/atlasent/audit_bundle.py
"""Offline verification for audit export bundles.

Mirrors ``atlasent-api/supabase/functions/v1-audit/verify.ts``. The
reference verifier is the source of truth; this file must stay
byte-identical with it on the canonicalization + signing path so a
bundle that verifies in the backend verifies here (and vice versa).

Public entry point:

    from atlasent import verify_bundle
    result = verify_bundle("export.json", public_keys_pem=[pem])

``public_keys_pem`` takes one or more SPKI-PEM strings — the active
set from ``GET /v1-signing-keys``. When omitted, verification still
runs the chain integrity check but ``signature_valid`` is False with
a descriptive ``reason``.
"""

from __future__ import annotations

import hashlib
import json
from base64 import urlsafe_b64decode
from collections.abc import Iterable
from dataclasses import dataclass, field
from pathlib import Path
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey


def _require_crypto():  # type: ignore[return]
    """Import cryptography at call-time; raise a clear error if absent."""
    try:
        from cryptography.exceptions import InvalidSignature  # noqa: PLC0415
        from cryptography.hazmat.primitives import serialization  # noqa: PLC0415
        from cryptography.hazmat.primitives.asymmetric.ed25519 import (  # noqa: PLC0415
            Ed25519PublicKey,
        )
        return InvalidSignature, serialization, Ed25519PublicKey
    except ImportError as exc:
        raise ImportError(
            "atlasent[verify] is required for audit-bundle verification. "
            "Install it with: pip install 'atlasent[verify]'"
        ) from exc

GENESIS_HASH = "0" * 64


@dataclass
class VerifyKey:
    """A public key the verifier will try, tagged with its registry id."""

    key_id: str
    public_key: Ed25519PublicKey


@dataclass
class BundleVerificationResult:
    """Outcome of verifying an audit export bundle.

    ``chain_integrity_ok`` is the AND of three checks: adjacency (each
    event's ``previous_hash`` equals the prior event's ``hash``),
    per-event hash recomputation from the canonical payload, and
    ``chain_head_hash`` matching the last event's stored hash.
    ``signature_valid`` is independent — a tampered bundle can have a
    valid signature over its original bytes, and vice versa.
    """

    chain_integrity_ok: bool
    signature_valid: bool
    head_hash_matches: bool
    tampered_event_ids: list[str] = field(default_factory=list)
    matched_key_id: str | None = None
    reason: str | None = None

    @property
    def verified(self) -> bool:
        """Convenience: everything passed."""

        return self.chain_integrity_ok and self.signature_valid


# ─── Canonicalization ────────────────────────────────────────────────


def canonical_json(value: Any) -> str:
    """RFC 8785-ish canonical JSON.

    Reproduces ``_shared/rules.ts::canonicalJSON`` byte-for-byte:
      * object keys sorted at every depth
      * no whitespace
      * ``None``, ``NaN``, ``+inf``, ``-inf`` all render as ``null``
      * strings use the same escapes as ``json.dumps`` with
        ``ensure_ascii=False``
    """

    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, int):
        return json.dumps(value)
    if isinstance(value, float):
        if value != value or value in (float("inf"), float("-inf")):
            return "null"
        return json.dumps(value)
    if isinstance(value, str):
        return json.dumps(value, ensure_ascii=False)
    if isinstance(value, list):
        return "[" + ",".join(canonical_json(v) for v in value) + "]"
    if isinstance(value, dict):
        parts = []
        for k in sorted(value.keys()):
            parts.append(
                json.dumps(k, ensure_ascii=False) + ":" + canonical_json(value[k])
            )
        return "{" + ",".join(parts) + "}"
    raise TypeError(f"cannot canonicalize {type(value).__name__}")


def _sha256_hex(s: str) -> str:
    return hashlib.sha256(s.encode("utf-8")).hexdigest()


# ─── Envelope reconstruction ─────────────────────────────────────────


def signed_bytes_for(bundle: dict) -> bytes:
    """Recreate the exact bytes ``handleExport`` signed.

    Key order is load-bearing — must match the object literal in
    ``v1-audit/index.ts::handleExport``. Python dicts preserve
    insertion order since 3.7, so the explicit construction below is
    the same wire output as V8/Deno's ``JSON.stringify``.
    """

    envelope = {
        "export_id": bundle.get("export_id"),
        "org_id": bundle.get("org_id"),
        "chain_head_hash": bundle.get("chain_head_hash"),
        "event_count": bundle.get("event_count"),
        "signed_at": bundle.get("signed_at"),
        "events": bundle.get("events"),
    }
    return json.dumps(envelope, separators=(",", ":"), ensure_ascii=False).encode(
        "utf-8"
    )


# ─── Chain verification ──────────────────────────────────────────────


def _verify_chain(events: list[dict]) -> tuple[bool, list[str]]:
    tampered: list[str] = []
    adjacency_ok = True
    prev_hash = events[0].get("previous_hash", GENESIS_HASH) if events else GENESIS_HASH
    if not isinstance(prev_hash, str):
        prev_hash = GENESIS_HASH

    for i, e in enumerate(events):
        stored_hash = e.get("hash")
        stored_prev = e.get("previous_hash")
        if not isinstance(stored_hash, str) or not isinstance(stored_prev, str):
            tampered.append(str(e.get("id", f"index_{i}")))
            adjacency_ok = False
            continue
        if stored_prev != prev_hash:
            adjacency_ok = False

        canonical = canonical_json(e.get("payload") or {})
        recomputed = _sha256_hex(prev_hash + canonical)
        if recomputed != stored_hash:
            tampered.append(str(e.get("id")))

        prev_hash = stored_hash

    return adjacency_ok, tampered


# ─── Signature verification ──────────────────────────────────────────


def _b64url_decode(s: str) -> bytes:
    pad = "=" * (-len(s) % 4)
    return urlsafe_b64decode(s + pad)


def _load_keys(public_keys_pem: Iterable[str] | None) -> list[VerifyKey]:
    if public_keys_pem is None:
        return []
    _, serialization, Ed25519PublicKey = _require_crypto()
    out: list[VerifyKey] = []
    for i, pem in enumerate(public_keys_pem):
        try:
            key = serialization.load_pem_public_key(pem.encode("utf-8"))
        except ValueError:
            continue
        if isinstance(key, Ed25519PublicKey):
            out.append(VerifyKey(key_id=f"pem_{i}", public_key=key))
    return out


def verify_audit_bundle(
    bundle: dict,
    keys: list[VerifyKey],
) -> BundleVerificationResult:
    """Verify a parsed bundle dict against a set of candidate public keys."""

    events = bundle.get("events") or []
    if not isinstance(events, list):
        events = []

    adjacency_ok, tampered = _verify_chain(events)

    if events and isinstance(events[-1].get("hash"), str):
        last_hash = events[-1]["hash"]
    else:
        last_hash = GENESIS_HASH
    head_hash_matches = bundle.get("chain_head_hash") == last_hash

    chain_integrity_ok = adjacency_ok and not tampered and head_hash_matches

    signature_valid = False
    matched_key_id: str | None = None
    reason: str | None = None

    signature = bundle.get("signature")
    if not keys:
        reason = (
            "no signing keys configured (signing_keys table empty and "
            "ATLASENT_EXPORT_SIGNING_KEY_PUBLIC unset)"
        )
    elif not isinstance(signature, str) or not signature:
        reason = "bundle carries no signature"
    else:
        try:
            sig_bytes = _b64url_decode(signature)
            envelope = signed_bytes_for(bundle)
            hint = (
                bundle.get("signing_key_id")
                if isinstance(bundle.get("signing_key_id"), str)
                else None
            )
            ordered = (
                [k for k in keys if k.key_id == hint]
                + [k for k in keys if k.key_id != hint]
                if hint
                else keys
            )
            InvalidSignature, _, __ = _require_crypto()
            for k in ordered:
                try:
                    k.public_key.verify(sig_bytes, envelope)
                except InvalidSignature:
                    continue
                signature_valid = True
                matched_key_id = k.key_id
                break
            if not signature_valid:
                reason = (
                    f"signature did not verify under any of "
                    f"{len(keys)} configured public key(s)"
                )
        except (ValueError, TypeError) as err:
            reason = f"signature check failed: {err}"

    if not chain_integrity_ok and reason is None:
        if tampered:
            reason = f"hash mismatch for {len(tampered)} event(s)"
        elif not adjacency_ok:
            reason = "chain adjacency broken"
        elif not head_hash_matches:
            reason = "chain_head_hash does not match last event"

    return BundleVerificationResult(
        chain_integrity_ok=chain_integrity_ok,
        signature_valid=signature_valid,
        head_hash_matches=head_hash_matches,
        tampered_event_ids=tampered,
        matched_key_id=matched_key_id,
        reason=reason,
    )


def verify_bundle(
    path: str | Path,
    public_keys_pem: Iterable[str] | None = None,
) -> BundleVerificationResult:
    """Load a JSON bundle from disk and verify it.

    ``public_keys_pem`` is the set of active SPKI-PEM strings from
    ``GET /v1-signing-keys``. When omitted, the chain check still runs
    but ``signature_valid`` is always False (with an explanatory
    ``reason``) — so callers that want a complete offline check MUST
    supply the trust set.
    """

    data = json.loads(Path(path).read_text(encoding="utf-8"))
    # Some fixtures wrap the bundle: {"description": "...", "bundle": {...}}.
    # Both shapes are accepted for developer ergonomics.
    if isinstance(data, dict) and "bundle" in data and isinstance(data["bundle"], dict):
        data = data["bundle"]

    keys = _load_keys(public_keys_pem)
    return verify_audit_bundle(data, keys)
